### PR TITLE
Accept `Duration` for all time-based combinators

### PR DESCRIPTION
**Pull request by armanbilge**
Follow-up to https://github.com/typelevel/cats-effect/pull/2954.

### CODE REVIEW

File: kernel/shared/src/main/scala/cats/effect/kernel/GenTemporal.scala [hunk 1]
```diff
@@ -278,13 +287,15 @@ object GenTemporal {
...
+      ifFinite: FiniteDuration => A): A =
     duration match {
-      case _: Duration.Infinite => fa
-      case finite: FiniteDuration => f(finite)
+      case fd: FiniteDuration => ifFinite(fd)
+      case Duration.Inf => ifInf
+      case d =>
+        throw new IllegalArgumentException(
+          s"Duration must be either a `FiniteDuration` or `Duration.Inf`, but got: $d")
     }
```
Comment (armanbilge) [inline kernel/shared/src/main/scala/cats/effect/kernel/GenTemporal.scala:298]: While we're here, should we also do something about negative `FiniteDuration`s?

**Comment by armanbilge** [comment]
Something I'm leery about is that these changes introduce a lot of branches that aren't currently tested by our laws which I think work in terms of `FiniteDuration`. We should change the laws as well, and if not then at least add specific test coverage for this.

**Review by djspiewak — DISMISSED** [review]
This seems good if we can get it to build. :-)

**Comment by armanbilge** [comment]
> This seems good if we can get it to build. :-)

Heh, I was waiting for a positive sign before continuing :) I'm on it!

**Pull request merged**